docs/docs/how-tos/encryption/custom_encryption.md

```diff
@@ -0,0 +1,451 @@
+            # Encrypt the VALUE for customer-specific sensitive fields
+            response = kms_client.encrypt(
+                KeyId=key_id,
+                Plaintext=str(value).encode(),
+                EncryptionContext={'tenant_id': tenant_id}
+            )
+            import base64
+            encrypted_data[key] = base64.b64encode(
+                response['CiphertextBlob']
+            ).decode()
+        else:
+            # Pass through unencrypted (including "owner" for search)
+            encrypted_data[key] = value
+
+    return encrypted_data
+
+
+@encrypt.decrypt.json
+async def decrypt_json_data(ctx: EncryptionContext, data: dict) -> dict:
+    """Decrypt JSON data fields.
+
+    Inverse of `encrypt_json_data`. Decrypts VALUES for fields with
+    "my.customer.org/" prefix, passes through all other fields.
+
+    The encryption context (tenant_id, key_id) is automatically retrieved
+    from storage, so callers don't need to re-pass the X-Encryption-Context header.
+
+    Args:
+        ctx: Encryption context with stored metadata
+        data: The encrypted data dictionary
+
+    Returns:
+        Decrypted data dictionary
+    """
+    tenant_id = ctx.metadata.get("tenant_id", "default")
+
+    decrypted_data = {}
+    for key, value in data.items():
+        if key.startswith("my.customer.org/") and isinstance(value, str):
+            # Decrypt the VALUE for customer-specific fields
+            import base64
+            ciphertext = base64.b64decode(value.encode())
+            response = kms_client.decrypt(
+                CiphertextBlob=ciphertext,
+                EncryptionContext={'tenant_id': tenant_id}
+            )
+            decrypted_data[key] = response['Plaintext'].decode()
+        else:
+            # Pass through unencrypted fields
+            decrypted_data[key] = value
+
+    return decrypted_data
+```
+
+### 2. Configure in langgraph.json
+
+Add the path to your encryption module in your `langgraph.json`:
+
+```json hl_lines="7-9"
+{
+  "dependencies": ["."],
+  "graphs": {
+    "agent": "./agent.py:graph"
+  },
+  "env": ".env",
+  "encrypt": {
+    "path": "./encrypt.py:encrypt"
+  }
+}
+```
+
+### 3. Pass encryption context in requests
+
+When making requests to your deployment, include the `X-Encryption-Context` header with a base64-encoded JSON object containing your encryption parameters:
+
+=== "Python Client"
+
+    ```python
+    import base64
+    import json
+    from langgraph_sdk import get_client
+
+    # Define encryption context
+    encryption_context = {
+        "tenant_id": "customer-123",
+        "key_id": "arn:aws:kms:us-east-1:123456789:key/abc-def-123"
+    }
+
+    # Encode as base64
+    encoded_context = base64.b64encode(
+        json.dumps(encryption_context).encode()
+    ).decode()
+
+    # Create client with encryption context header
+    client = get_client(
+        url="http://localhost:2024",
+        headers={"X-Encryption-Context": encoded_context}
```

Comment:
This client always provides the same value for the encryption context header. In practice, we need to dynamically pass different tenant IDs/key IDs (in our case it is the same) per request. We already created our own SDK clients that provide a dynamic JWT header per request which injects different values for the brand ID (aka tenant ID). This is because we invoke the same graph for different tenants and need to reuse the client in a way that each API can receive different values for those headers (instead of being always the same during initialization)